tests/genai/utils/test_trace_utils.py

```diff
@@ -955,3 +958,311 @@ async def slow_async_predict_fn(request):
         converted_fn(request=sample_input)
 
     assert len(get_traces()) == 0
+
+
+@pytest.mark.parametrize(
+    ("span_type", "use_attribute", "tool_name", "tool_description"),
+    [
+        ("LLM", True, "get_weather", "Get current weather"),
+        ("CHAT_MODEL", False, "search", "Search the web"),
+    ],
+)
+def test_extract_available_tools_from_trace_basic(
+    span_type, use_attribute, tool_name, tool_description
+):
+    tools = [
+        {
+            "type": "function",
+            "function": {
+                "name": tool_name,
+                "description": tool_description,
+                "parameters": {"type": "object", "properties": {"param": {"type": "string"}}},
+            },
+        }
+    ]
+
+    with mlflow.start_span(name="test_span", span_type=span_type) as span:
+        if use_attribute:
+            set_span_chat_tools(span, tools)
+            span.set_inputs({"prompt": "test"})
+        else:
+            span.set_inputs({"messages": [{"role": "user", "content": "test"}], "tools": tools})
+        span.set_outputs({"response": "result"})
+
+    trace = mlflow.get_trace(span.trace_id)
+    extracted_tools = extract_available_tools_from_trace(trace)
+
+    assert len(extracted_tools) == 1
+    assert extracted_tools[0].function.name == tool_name
+    assert extracted_tools[0].function.description == tool_description
+
+
+def test_extract_available_tools_from_trace_with_multiple_spans():
+    tool1 = [
+        {
+            "type": "function",
+            "function": {
+                "name": "add",
+                "description": "Add two numbers",
+                "parameters": {
+                    "type": "object",
+                    "properties": {
+                        "a": {"type": "number"},
+                        "b": {"type": "number"},
+                    },
+                },
+            },
+        }
+    ]
+
+    tool2 = [
+        {
+            "type": "function",
+            "function": {
+                "name": "multiply",
+                "description": "Multiply two numbers",
+                "parameters": {
+                    "type": "object",
+                    "properties": {
+                        "x": {"type": "number"},
+                        "y": {"type": "number"},
+                    },
+                },
+            },
+        }
+    ]
+
+    with mlflow.start_span(name="parent") as parent:
+        with mlflow.start_span(name="llm1", span_type="LLM") as span1:
+            set_span_chat_tools(span1, tool1)
+
+        with mlflow.start_span(name="llm2", span_type="CHAT_MODEL") as span2:
+            set_span_chat_tools(span2, tool2)
+
+    trace = mlflow.get_trace(parent.trace_id)
+    extracted_tools = extract_available_tools_from_trace(trace)
+
+    assert len(extracted_tools) == 2
+    tool_names = [t.function.name for t in extracted_tools]
+    assert "add" in tool_names
```

Comment:
nit: can we assert on the exact format of extracted tools? this is generally safer than asserting on a portion of `extracted_tools`